src/Symfony/Component/String/Utf8String.php

```diff
@@ -0,0 +1,249 @@
+            $needle = $needle->string;
+        } elseif (!\is_string($needle)) {
+            return parent::indexOfLast($needle, $offset);
+        }
+
+        if ('' === $needle) {
+            return null;
+        }
+
+        $i = $this->ignoreCase ? mb_strripos($this->string, $needle, $offset, 'UTF-8') : mb_strrpos($this->string, $needle, $offset, 'UTF-8');
+
+        return false === $i ? null : $i;
+    }
+
+    public function length(): int
+    {
+        return mb_strlen($this->string, 'UTF-8');
+    }
+
+    public function prepend(string ...$prefix): AbstractString
+    {
+        $str = clone $this;
+        $str->string = (1 >= \count($prefix) ? $prefix[0] ?? '' : implode('', $prefix)).$this->string;
+
+        if (!preg_match('//u', $str->string)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        return $str;
+    }
+
+    public function replace(string $from, string $to): AbstractString
+    {
+        $str = clone $this;
+
+        if ('' === $from || !preg_match('//u', $from)) {
+            return $str;
+        }
+
+        if ('' !== $to && !preg_match('//u', $to)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        if ($this->ignoreCase) {
+            $str->string = implode($to, preg_split('{'.preg_quote($from).'}iu', $this->string));
+        } else {
+            $str->string = str_replace($from, $to, $this->string);
+        }
+
+        return $str;
+    }
+
+    public function slice(int $start = 0, int $length = null): AbstractString
+    {
+        $str = clone $this;
+        $str->string = (string) mb_substr($this->string, $start, $length, 'UTF-8');
+
+        return $str;
+    }
+
+    public function splice(string $replacement, int $start = 0, int $length = null): AbstractString
+    {
+        if (!preg_match('//u', $replacement)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $str = clone $this;
+        $start = $start ? \strlen(mb_substr($this->string, 0, $start, 'UTF-8')) : 0;
+        $length = $length ? \strlen(mb_substr($this->string, $start, $length, 'UTF-8')) : $length;
+        $str->string = substr_replace($this->string, $replacement, $start, $length ?? \PHP_INT_MAX);
+
+        return $str;
+    }
+
+    public function split(string $delimiter, int $limit = null, int $flags = null): array
+    {
+        if (1 > $limit = $limit ?? \PHP_INT_MAX) {
+            throw new InvalidArgumentException('Split limit must a positive integer.');
+        }
+
+        if ('' === $delimiter) {
+            throw new InvalidArgumentException('Split delimiter is empty.');
+        }
+
+        if (null !== $flags) {
+            return parent::split($delimiter.'u', $limit, $flags);
+        }
+
+        if (!preg_match('//u', $delimiter)) {
+            throw new InvalidArgumentException('Split delimiter is invalid UTF-8 string.');
```

Comment:
IMO it should be `is not valid` instead of `is invalid` (would fit better with condition that checks if does NOT match regex). But it's subjective :slightly_smiling_face:

Reply:
accepted: `Split delimiter is not a valid UTF-8 string.`